dali/kernels/imgproc/pointwise/multiply_add.h

```diff
@@ -15,25 +15,81 @@
 #ifndef DALI_KERNELS_IMGPROC_POINTWISE_MULTIPLY_ADD_H_
 #define DALI_KERNELS_IMGPROC_POINTWISE_MULTIPLY_ADD_H_
 
+#include <limits>
+#include <tuple>
 #include <utility>
-#include "dali/util/ocv.h"
 #include "dali/core/common.h"
 #include "dali/core/convert.h"
-#include "dali/core/geom/box.h"
 #include "dali/core/error_handling.h"
-#include "dali/kernels/kernel.h"
+#include "dali/core/geom/box.h"
 #include "dali/kernels/imgproc/roi.h"
+#include "dali/kernels/kernel.h"
+#include "dali/util/ocv.h"
 
 namespace dali {
 namespace kernels {
 
+template <typename T, typename... TS>
+struct AnyOf {
+  static constexpr bool value = (std::is_same_v<T, TS> || ...);
+};
+
+template <typename T>
+struct MakeUnsigned {
+  using type = std::make_unsigned_t<T>;
+};
+
+template <>
+struct MakeUnsigned<bool> {
+  using type = bool;
+};
+
+template <typename Out, typename In, bool use_lut = AnyOf<In, uint8_t, int8_t, bool>::value>
+struct MultiplyAddElementCpu;
+
+template <typename Out, typename In>
+struct MultiplyAddElementCpu<Out, In, false> {
+  MultiplyAddElementCpu(float addend, float multiplier)
+      : addend_{addend}, multiplier_{multiplier} {}
+
+  Out operator()(In element) {
+    return ConvertSat<Out>(element * multiplier_ + addend_);
+  }
+
+  float addend_;
+  float multiplier_;
+};
+
+template <typename Out, typename In>
+struct MultiplyAddElementCpu<Out, In, true> {
+  using InUnsigned = typename MakeUnsigned<In>::type;
+  static constexpr int kRangeSize = std::numeric_limits<InUnsigned>::max() + 1;
+  // if that's deliberate you can always lift the restriction
+  static_assert(kRangeSize <= 256, "Using lut for bigger types may not be a good idea");
+
+  MultiplyAddElementCpu(float addend, float multiplier) : addend_{addend}, multiplier_{multiplier} {
+    for (int val = std::numeric_limits<In>::min(); val <= std::numeric_limits<In>::max(); val++) {
+      lut_[static_cast<InUnsigned>(val)] = ConvertSat<Out>(val * multiplier_ + addend_);
+    }
+  }
+
+  Out operator()(InUnsigned element) {
+    return lut_[element];
+  }
```

Comment:
```suggestion
  Out operator()(In element) {
    return lut_[static_cast<Unsigned>(element)];
  }
```
This will let you avoid some potentially very nasty surprises (and possibly warnings). We know what we're doing inside this function, but the user doesn't need to see this trick.

Reply:
done